modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/MajorityPeerSelect.scala

```diff
@@ -50,28 +60,34 @@ object MajorityPeerSelect {
 
     def filterPeerList(peers: NonEmptyList[Peer]): F[NonEmptyList[Peer]] =
       peers
-        .traverse(snapshotClient.getLatestOrdinal(_))
-        .map {
-          _.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size }
+        .parTraverseN(maxConcurrentPeerInquiries)(snapshotClient.getLatestOrdinal(_))
+        .map(_.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size })
+        .map { case (majorityOrdinal, _) => majorityOrdinal }
+        .flatMap { majorityOrdinal =>
+          peers
+            .parTraverseN(maxConcurrentPeerInquiries)(peer => getPeerHash(peer, majorityOrdinal).attempt)
         }
-        .flatMap {
-          case (majorityOrdinal, _) =>
-            peers.traverse(snapshotClient.get(majorityOrdinal).run(_).flatMap(_.toHashed.map(_.hash)))
+        .flatMap { maybePeerSnapshots =>
+          MonadThrow[F].fromOption(
+            maybePeerSnapshots.collect { case Right(peerSnapshot) => peerSnapshot }.toNel,
+            NoPeersToSelect
+          )
         }
-        .map(_.zip(peers))
-        .map(_.groupMap { case (hash, _) => hash } { case (_, ps) => ps })
-        .map(_.maxBy { case (_, peers) => peers.size })
-        .map { case (_, peerCandidates) => peerCandidates }
+        .map(_.groupMap { case (_, hash) => hash } { case (peer, _) => peer })
+        .map(_.values.toList.sortWith(_.size > _.size))
+        .map(_.head)
 
     def getPeerSublist(peers: Set[Peer]): F[List[Peer]] = {
-      val maxSublistPercent = 0.25
-      val maxSublistSize = Math.max((peers.size * maxSublistPercent).toInt, 1)
+      val maxSampleSize = minPeerSelectionSize + (peers.size * peerSampleRatio).toInt
 
-      Random[F].nextIntBounded(maxSublistSize).flatMap { peerCount =>
+      Random[F].betweenInt(minPeerSelectionSize, maxSampleSize + 1).flatMap { peerCount =>
         Random[F]
           .shuffleList(peers.toList)
           .map(_.take(peerCount))
       }
```

Comment:
Here are a couple alternate implementations which IMO are cleaner:
```
      for {
        shuffled <- Random[F].shuffleList(peers.toList)
        takeAmount <- Random[F].betweenInt(minPeerSelectionSize, maxSampleSize + 1)
      } yield shuffled.take(takeAmount)
```
```
      Applicative[F].map2(
        Random[F].shuffleList(peers.toList),
        Random[F].betweenInt(minPeerSelectionSize, maxSampleSize + 1)
      )((ps, n) => ps.take(n))
```

Reply:
I'd see it as:
```scala
val sampleSize = (peers.size * peerSampleRatio).toInt

Random[F].shuffleList(peers.toList)
  .map(_.take(Math.max(sampleSize, minSampleSize)))
```

so we either take 25% of the cluster or all the peers (if the sample size is lower than min sample size)

Reply:
I think that implementation shortcoming is fine, @marcinwadon. I'll check with @ryle-goehausen to be sure.

Reply:
Implemented @marcinwadon suggestion.